### DISCUSSION TITLE
Allow pushing multiple tags to BSR

### DISCUSSION THREAD

**Discussion by rohilsurana** [bufbuild/buf#576] [Ideas]
As of now only the commit hash is pushed to BSR as a tag.
It's not possible to push a git tag as BSR tag or push both as tags using this action. Basically the equivalent of -

`buf push --tag "v0.0.5" --tag "$(git rev-parse HEAD)"`

It would be good to have these configurable to push commit hash, git tag, and/or git branch.

**Comment by amckinney** [comment]
Thanks for the issue! There are a couple things to consider here, so I want to separate them into two different discussions.

## Tags
How do you imagine this being configured? Given that a single git commit can have multiple tags, would you want `buf` to tag the BSR commit with _all_ of the tags present on the git commit? For example, consider the following commit with two tags:

```sh
$ git log
...
commit 7a342897db48fc51c4225e83af3f7f54173b7cbb (tag: v0.0.24-staging, tag: v0.0.24)
...
```

Theoretically, we could expose a configuration in the [action.yml](https://github.com/bufbuild/buf-push-action/blob/040008e05aecab7de4f5237a341c16bba68eba42/action.yml#L7) with something like the following:

```yaml
inputs:
  include_git_tags:
    description: 'If enabled, create a BSR tag for each git tag present on the current git commit.'
    default: false
```

However, this makes it difficult to configure behavior with finer granularity. What if I only want certain tags and not others (i.e. `v0.0.24` and not `v0.0.24-staging`)? It's not practical to select specific tags on a commit-by-commit basis, since tags _must_ change between commits (for uniqueness), so you would need to reconfigure your action every time.

I'm not necessarily opposed to this idea, but I want to make sure I understand your use case, and that pushing _all_ the tags is what you expect to see and would be beneficial to you (especially given the caveats I mentioned above).

## Branches
As for your note about git branch support, we're considering our own concept of BSR branches that would take the git branch into account when pushing (i.e. pushing to the `dev` branch in your git repository would similarly push to the `dev` branch in the BSR). We haven't yet finalized this decision, but it's something we're actively considering.

Without BSR branches, tagging by git branches is otherwise difficult since branch names are frequently reused (e.g. `dev`) and could frequently trigger uniqueness constraint errors when we try to push.

**Comment by rohilsurana** [comment]
I was thinking of a use case where we have triggers only on tag creation.
I was thinking of giving users a way to specify custom multiple tags in inputs. User could use stuff from `env` context to populate these.

**Comment by rohilsurana** [comment]
As for branches, I think what you are saying makes sense. Will keep an eye out for the BSR concept of branch. 😄

**Comment by myles-mcdonnell** [comment]
Related to this discussion...

In our case we have a large GRPC micro service estate.  We publish versioned (semver) services.  The version exists as a tag on the source code and a matching tag on the deployment asset (docker image).  When integrating buf I was expecting to be able to build a pipeline that also publishes a schema with the same tag.  This seems like an essential schema registry capability to me.  Maybe this is possible and I just haven't worked out how yet?
